hw/ip/otbn/dv/otbnsim/sim/sim.py

```diff
@@ -44,29 +44,28 @@ def step(self, verbose: bool) -> Tuple[Optional[OTBNInsn], List[Trace]]:
         if not self.state.running:
             return (None, [])
 
-        was_stalled = self.state.stalled
         pc_before = self.state.pc
 
-        if was_stalled:
+        word_pc = int(self.state.pc) >> 2
+        if word_pc >= len(self.program):
+            raise RuntimeError('Trying to execute instruction at address '
+                               '{:#x}, but the program is only {:#x} '
+                               'bytes ({} instructions) long. Since there '
+                               'are no architectural contents of the '
+                               'memory here, we have to stop.'
+                               .format(int(self.state.pc),
+                                       4 * len(self.program),
+                                       len(self.program)))
+        insn = self.program[word_pc]
+
+        sim_stalled = (self.state.non_insn_stall or
+                       not insn.pre_execute(self.state))
```

Comment:
This is correct, but could you split it up to something like this?
```
sim_stalled = self.state.non_insn_stall
if not sim_stalled:
  sim_stalled = not insn.pre_execute(self.state)
```
Since `insn.pre_execute` can have side-effects, I think it's probably a bit easier to read if you don't need to remember about `or`'s short-circuiting behaviour.